File: packages/mqtt_topics/topics_loader.ts
import mqttTopics from './topics.json';

// TopicConfig interface matching Python TopicConfig
export interface TopicConfig {
  topic: string;
  qos: number;
  retain: boolean;
}

// Type for the structure of topics.json with replaced values
export type AllTopics = {
  command: {
    get_config: {
      request: TopicConfig;
      response: TopicConfig;
    };
    reboot: {
      request: TopicConfig;
      response: TopicConfig;
    };
    set_value_to_point: {
      request: TopicConfig;
      response: TopicConfig;
    };
    start_monitoring: {
      request: TopicConfig;
      response: TopicConfig;
    };
    stop_monitoring: {
      request: TopicConfig;
      response: TopicConfig;
    };
  };
  status: {
    heartbeat: TopicConfig;
  };
  data: {
    point: TopicConfig | null;
    point_bulk: TopicConfig;
  };
  alert_management: {
    acknowledge: TopicConfig;
    resolve: TopicConfig;
  };
};

// Strongly typed params for all possible placeholders in topics.json
export type TopicParams = {
  organization_id?: string;
  site_id?: string;
  iot_device_id?: string;
  controller_device_id?: string;
  iot_device_point_id?: string;
  tenant_id?: string;
};

export enum CommandNameEnum {
  GET_CONFIG = 'get_config',
  REBOOT = 'reboot',
  SET_VALUE_TO_POINT = 'set_value_to_point',
  START_MONITORING = 'start_monitoring',
  STOP_MONITORING = 'stop_monitoring',
}

// MQTT message payload types matching Python ControllerPointsModel

export interface ControllerPoint {
  id: number | null
  controller_ip_address: string
  controller_port: number
  bacnet_object_type: string
  point_id: number
  iot_device_point_id: string
  controller_id: string
  units: string | null
  present_value: string | null
  controller_device_id: string
  created_at: string
  updated_at: string
  created_at_unix_milli_timestamp: number

  // Health monitoring fields
  status_flags: string[] | null
  event_state: string | null
  out_of_service: boolean | null
  reliability: string | null

  // Value limit properties
  min_pres_value: number | null
  max_pres_value: number | null
  high_limit: number | null
  low_limit: number | null
  resolution: number | null

  // Control properties
  priority_array: unknown | null
  relinquish_default: number | null

  // Notification configuration properties
  cov_increment: number | null
  time_delay: number | null
  time_delay_normal: number | null
  notification_class: number | null
  notify_type: string | null
  deadband: number | null
  limit_enable: unknown | null

  // Event properties
  event_enable: unknown | null
  acked_transitions: unknown | null
  event_time_stamps: unknown | null
  event_message_texts: unknown | null
  event_message_texts_config: unknown | null

  // Algorithm control properties
  event_detection_enable: boolean | null
  event_algorithm_inhibit_ref: unknown | null
  event_algorithm_inhibit: boolean | null
  reliability_evaluation_inhibit: boolean | null

  error_info: string | null
}

export interface PointBulkPayload {
  points: ControllerPoint[]
}

function buildTopic(template: string, params: Readonly<Record<string, string>>): string {
  return template.replace(/\{(\w+)\}/g, (_m, key: string) =>
    Object.prototype.hasOwnProperty.call(params, key) && params[key] !== undefined
      ? params[key]
      : `{${key}}`
  );
}

export function getAllTopics({ params }: { params: TopicParams }): AllTopics {
  function replaceDeep<T>(node: T): T {
    if (typeof node === 'string') {
      return buildTopic(node as unknown as string, params as Record<string, string>) as unknown as T;
    }
    if (node && typeof node === 'object') {
      if (Array.isArray(node)) {
        return (node as unknown[]).map((n) => replaceDeep(n)) as unknown as T;
      }
      const next: Record<string, unknown> = {};
      for (const [k, v] of Object.entries(node as Record<string, unknown>)) {
        next[k] = replaceDeep(v);
      }
      return next as T;
    }
    return node;
  }

  const replaced = replaceDeep<typeof mqttTopics>(mqttTopics);
  const missingPointIds = !params.controller_device_id || !params.iot_device_point_id;
  return {
    ...replaced,
    data: {
      ...replaced.data,
      point: missingPointIds ? null : replaced.data.point,
    },
  } as AllTopics;
}

function buildMQTTSubscriptionPattern(topicTemplate: string): string {
  return topicTemplate.replace(/\{[^}]+\}/g, '+');
}

export const getGlobalTopicsToWriteToDB = (): string[] => {
  // MQTT shared subscription with queue for load balancing across multiple instances
  // See: https://docs.emqx.com/en/emqx/latest/messaging/mqtt-shared-subscription.html
  // Using $queue pattern to ensure only one instance processes each message

  const topics = [
    buildMQTTSubscriptionPattern(mqttTopics.data.point_bulk.topic),
    buildMQTTSubscriptionPattern(mqttTopics.command.set_value_to_point.response.topic).replace('set_value_to_point', '+'),
    buildMQTTSubscriptionPattern(mqttTopics.status.heartbeat.topic),
    buildMQTTSubscriptionPattern(mqttTopics.alert_management.acknowledge.topic),
    buildMQTTSubscriptionPattern(mqttTopics.alert_management.resolve.topic)
  ];

  // Add queue prefix for load balancing across multiple instances
  return topics.map((topic) => `$queue/${topic}`);
}

File: packages/mqtt_topics/topics_loader.py
import json
from pathlib import Path
from typing import Dict, Any, List, Tuple, Optional
from typing import TypedDict, cast
from pydantic import BaseModel
from enum import Enum

# Add more as your schema grows

class TopicConfig(BaseModel):
    topic: str
    qos: int = 1
    retain: bool = False

class CommandEntry(BaseModel):
    request: TopicConfig
    response: TopicConfig

class CommandSection(BaseModel):
    get_config: CommandEntry
    reboot: CommandEntry
    set_value_to_point: CommandEntry
    start_monitoring: CommandEntry
    stop_monitoring: CommandEntry
    # Add more commands as your schema grows

class StatusSection(BaseModel):
    heartbeat: TopicConfig

class DataSection(BaseModel):
    point: Optional[TopicConfig] = None
    point_bulk: Optional[TopicConfig] = None

class AlertManagementSection(BaseModel):
    acknowledge: TopicConfig
    resolve: TopicConfig

class Topics(BaseModel):
    command: CommandSection  # Always present
    status: StatusSection
    data: DataSection
    alert_management: AlertManagementSection

class CommandNameEnum(str, Enum):
    get_config = "get_config"
    reboot = "reboot"
    set_value_to_point = "set_value_to_point"
    start_monitoring = "start_monitoring"
    stop_monitoring = "stop_monitoring"
    # Add more commands as your schema grows

def load_mqtt_topics() -> Topics:
    mqtt_topics_path = Path(__file__).parent / 'topics.json'
    with open(mqtt_topics_path) as f:
        data = json.load(f)
    return Topics.model_validate(data)

def build_mqtt_topic_dict(
    organization_id: str,
    site_id: str,
    iot_device_id: str,
    controller_device_id: Optional[str] = None,
    iot_device_point_id: Optional[str] = None,
) -> Topics:
    """
    Build a full MQTT topic Topics model with placeholders replaced for the new topics.json schema, including controller_device_id and iot_device_point_id as optional.

    Args:
        organization_id (str): Organization identifier.
        site_id (str): Site identifier.
        iot_device_id (str): IoT device identifier.
        controller_device_id (Optional[str]): Controller device identifier.
        iot_device_point_id (Optional[str]): IoT device point identifier.

    Returns:
        Topics: A Topics Pydantic model with actual values filled in (or placeholders if not provided).
    """
    # Treat empty strings as None for optional fields
    if controller_device_id == "":
        controller_device_id = None
    if iot_device_point_id == "":
        iot_device_point_id = None
    # Validate that required values are not empty strings
    for name, value in {
        "organization_id": organization_id,
        "site_id": site_id,
        "iot_device_id": iot_device_id,
    }.items():
        if not value:
            raise ValueError(f"Missing placeholder value for {name}.")

    templates = load_mqtt_topics().model_dump()

    values = {
        "organization_id": organization_id,
        "site_id": site_id,
        "iot_device_id": iot_device_id,
    }
    if controller_device_id is not None:
        values["controller_device_id"] = controller_device_id
    if iot_device_point_id is not None:
        values["iot_device_point_id"] = iot_device_point_id

    def replace_placeholders(template_str: str) -> str:
        try:
            # Only replace placeholders for which we have values; leave others as is
            return template_str.format_map(Default(dict(values)))
        except KeyError as e:
            return template_str  # Leave the placeholder as is

    class Default(dict):
        def __missing__(self, key):
            return '{' + key + '}'

    def fill_templates(node: Any, parent_key: Optional[str] = None) -> Any:
        if isinstance(node, dict):
            result: dict[str, Any] = {}
            for key, value in node.items():
                # If we're in the 'data' section and the key is 'point', handle special logic
                if parent_key == 'data' and key == 'point':
                    if controller_device_id is None or iot_device_point_id is None:
                        result[key] = None
                    else:
                        result[key] = fill_templates(value, key)
                else:
                    result[key] = fill_templates(value, key)
            return result
        elif isinstance(node, str):
            return replace_placeholders(node)
        else:
            return node

    filled = fill_templates(templates)
    return Topics.model_validate(filled)

def build_mqtt_command_topic(
    organization_id: str,
    site_id: str,
    iot_device_id: str,
) -> CommandSection:
    """
    Build and return the CommandSection Pydantic model for the given device IDs.
    """
    topics = build_mqtt_topic_dict(
        organization_id=organization_id,
        site_id=site_id,
        iot_device_id=iot_device_id,
    )
    return topics.command


def build_mqtt_status_topic(
    organization_id: str,
    site_id: str,
    iot_device_id: str,
) -> StatusSection:
    """
    Build and return the StatusSection Pydantic model for the given device IDs.
    """
    topics = build_mqtt_topic_dict(
        organization_id=organization_id,
        site_id=site_id,
        iot_device_id=iot_device_id,
    )
    return topics.status


def build_mqtt_data_topic(
    organization_id: str,
    site_id: str,
    iot_device_id: str,
    controller_device_id: Optional[str] = None,
    iot_device_point_id: Optional[str] = None,
) -> DataSection:
    """
    Build and return the DataSection Pydantic model for the given device IDs.
    """
    topics = build_mqtt_topic_dict(
        organization_id=organization_id,
        site_id=site_id,
        iot_device_id=iot_device_id,
        controller_device_id=controller_device_id,
        iot_device_point_id=iot_device_point_id,
    )
    return topics.data


def build_mqtt_alert_management_topic(
    organization_id: str,
    site_id: str,
    iot_device_id: str,
) -> AlertManagementSection:
    """
    Build and return the AlertManagementSection Pydantic model for the given device IDs.
    """
    topics = build_mqtt_topic_dict(
        organization_id=organization_id,
        site_id=site_id,
        iot_device_id=iot_device_id,
    )
    return topics.alert_management
